### DISCUSSION TITLE
[FEATURE] FLUTE service

### DISCUSSION THREAD

**Discussion by apwiggins** [coreemu/core#243] [Ideas]
Is there interest in having a FLUTE service for CORE?  It requires compiling external packages, so this doesn't fit the pattern of using only OS-packaged binaries/configs.

https://tools.ietf.org/html/rfc6726

> FLUTE - File Delivery over Unidirectional Transport
> This document defines File Delivery over Unidirectional Transport (FLUTE), a protocol for the unidirectional delivery of files over the Internet, which is particularly suited to multicast networks.  The specification builds on Asynchronous Layered Coding, the base protocol designed for massively scalable multicast distribution.

Caution.  This service file hasn't been tested against CORE/master, so it *definitely* has issues.
https://gist.github.com/apwiggins/c1ed508d169d297e3d2307096a37e794

**Comment by b00ga** [comment]
CORE actually has no expectation of the required software being OS-packaged. In fact, CORE doesn't attempt to install any software for its services. That is expected of the end-user. CORE 4.x used to show every available service and the given service would just fail if it couldn't find the required binaries. CORE 5.x looks for the binaries defined in the CORE service, and if it can't find them, it disables that service (hiding it in the "Services" popup for a node).

In addition to adding services to upstream CORE, you can easily package/install your own CORE services outside of the upstream CORE project as well. Just add the python service file to a directory on disk and point to it with */etc/core/core.conf*.

Once you've run the CORE GUI once, you'll get a .core directory in your home directory. For example, let's say that's */home/apwiggins/.core* .  You can copy *flute.py* to */home/apwiggins/.core/myservices*, then edit */home/apwiggins/.core/myservices/\_\_init\_\_.py* and add "flute" to the **\_\_all\_\_** list, then edit */etc/core/core.conf*, add */home/apwiggins/.core/myservices/* to the **custom_services_dir** entry (there's an example entry commented out in the standard file) and restart the CORE daemon. Your flute service should appear in the GUI now.

**Comment by b00ga** [comment]
Actually just found the process I described in the following README: https://github.com/coreemu/core/blob/master/daemon/examples/myservices/README.txt

Taking a quick look at your gist, it looks like a CORE 4.x service. The "check for binaries" part I was speaking of in CORE 5.x requires you to define an "executable" instance variable in your service. You can see the latest sample CORE service in the repo here: https://github.com/coreemu/core/blob/master/daemon/examples/myservices/sample.py

**Comment by apwiggins** [comment]
b00ga... I'm well aware of how to add custom services.  In the past, the services add to master have been all?/mostly? based on OS-packaged binaries/configs.  The question is whether CORE/master wants to adopt additional community-offered services as part of the master tree.

It would be great if you could offer a community-offered DLEP service, for example.

**Comment by b00ga** [comment]
I guess I just disagree with the "OS-packaged binaries/configs" statement. There are a number of CORE services that are not available as OS-packaged binaries. I think a number of the underlying packages have been added to Ubuntu over the years, but when they became CORE services, they did not exist as Ubuntu packages. Also, there are a number of RHEL/CentOS CORE users out there, and a large quantity of master-tree CORE services software must be compiled from source on RHEL/CentOS, as there are no readily available packages.

I could talk to the DLEP developer about a CORE service. He may be interested. I don't have enough operational knowledge of DLEP to create a usable service.

It might be cool to have a community repository of CORE services. Here the community can vet them or enhance them, or maybe they don't make sense for upstream inclusion and it doesn't put the burden on the CORE developers to maintain services that they my not have enough experience with to support and maintain as part of master. Even something as simple as an "Awesome CORE repo" (check https://awesomerepos.com/ if you're unfamiliar with Awesome repos).

Anyway, sorry, I didn't realize you were as versed as you are with CORE services. Just trying to help.

**Comment by Kevlar** [comment]
We're looking to potentially create a second repository to support additional services in the near future.  Until that comes online, we're definitely open to additional services on the main repository (with the expectation that less "standard" services may get moved to the extended services repository) and would love contributions.
